apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/data_updater/impl.ex

```diff
@@ -1608,6 +1620,37 @@ defmodule Astarte.DataUpdaterPlant.DataUpdater.Impl do
     {:ack, :ok, final_state}
   end
 
+  def handle_capabilities(state, payload, _timestamp) do
+    %{device_id: device_id} = state
+
+    with {:ok, bson_payload} <- decode_bson_payload(state, payload),
+         changeset = Capabilities.changeset(state.capabilities, bson_payload),
+         {:ok, capabilities} <- update_capabilities(changeset, state) do
+      {:ok, db_client} = Database.connect(realm: state.realm)
+      Queries.set_device_capabilities(db_client, device_id, capabilities)
+
+      new_state = %State{state | capabilities: capabilities}
+      {:ack, :ok, new_state}
+    end
+  end
+
+  defp update_capabilities(changeset, state) do
+    with {:error, error} <- Ecto.Changeset.apply_action(changeset, :update) do
+      Logger.warning(
+        "Error while updating capabilities: #{error}",
+        tag: "capabilities_error"
+      )
+
+      :telemetry.execute(
+        [:astarte, :data_updater_plant, :data_updater, :discarded_message],
+        %{},
+        %{realm: state.realm}
+      )
+
+      {:discard, error, state}
```

Comment:
If the capabilities message sent by the device is malformed, Astarte should forcefully disconnect the device as it's being done e.g. for an [unexpected control message](https://github.com/astarte-platform/astarte/blob/7e10975abf83518f4ae7de52e1a6150bdf404f9a/apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/data_updater/impl.ex#L1719-L1728).
A possible option is to use an `:handle_continue` return value, as done e.g. for unexpected control messages (see [here](https://github.com/astarte-platform/astarte/blob/7e10975abf83518f4ae7de52e1a6150bdf404f9a/apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/data_updater/impl.ex#L1166))